### PR TITLE
[nereids] group by key elimination

### PR DESCRIPTION
**Pull request by xzj7019**
## Proposed changes

group by key elimination

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/ExprFdItem.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            ImmutableSet<SlotReference> childExprs) {
+        super(parentExprs, isUnique, false);
+        this.childExprs = ImmutableSet.copyOf(childExprs);
+    }
+
+    @Override
+    public boolean checkExprInChild(SlotReference slot, LogicalPlan childPlan) {
+        return childExprs.contains(slot);
+    }
+}
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/ExprFdItem.java:41]: The uniform information is lost?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/LogicalProperties.java [hunk 2]
```diff
@@ -42,11 +42,12 @@ public class LogicalProperties {
     protected final Supplier<Set<Slot>> outputSetSupplier;
     protected final Supplier<Map<Slot, Slot>> outputMapSupplier;
     protected final Supplier<Set<ExprId>> outputExprIdSetSupplier;
-    protected final Supplier<FunctionalDependencies> fdSupplier;
+    protected final Supplier<ImmutableSet<FdItem>> fdItemsSupplier;
     private Integer hashCode = null;
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/LogicalProperties.java:46]: why not put it in FunctionalDependencies and keep the same API with the origin

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupBy.java [hunk 3]
```diff
@@ -50,8 +53,11 @@ public Rule build() {
                     Set<Slot> groupby = agg.getGroupByExpressions().stream().map(e -> (Slot) e)
                             .collect(Collectors.toSet());
                     Plan child = agg.child();
-                    boolean unique = child.getLogicalProperties().getFunctionalDependencies()
-                            .isUniqueAndNotNull(groupby);
+                    //boolean unique = child.getLogicalProperties().getFunctionalDependencies()
+                    //        .isUniqueAndNotNull(groupby);
+                    ImmutableSet<FdItem> fdItems = child.getLogicalProperties().getFdItems();
+                    boolean unique = !groupby.isEmpty() && fdItems.stream().anyMatch(e ->
+                            e.isUnique() && !e.isCandidate() && groupby.containsAll(e.getParentExprs()));
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupBy.java:60]: It's incomplete. such as :
```
select * from (select a, b, sum(a + b) as c group by a, b) group by a b c
```
a c are unique, and unique a, b, c also can be infered . That is the uniqueness can be inferred by uniqueness of any subset

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java [hunk 4]
```diff
@@ -330,15 +331,17 @@ public static boolean canEliminateByFk(LogicalJoin<?, ?> join, Plan primaryPlan,
...
+    public static boolean canEliminateByLeft(LogicalJoin<?, ?> join, ImmutableSet<FdItem> fdItems) {
         if (join.getJoinType().isLeftOuterJoin()) {
             Pair<Set<Slot>, Set<Slot>> njHashKeys = join.extractNullRejectHashKeys();
             if (!join.getOtherJoinConjuncts().isEmpty() || njHashKeys == null) {
                 return false;
             }
-            return rightFuncDeps.isUnique(njHashKeys.second);
+            return fdItems.stream().anyMatch(e -> e.getParentExprs().containsAll(njHashKeys.second)
+                    && e.isUnique()
+                    && !e.isCandidate());
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:344]: as stated above, the inferring of uniqueness is incomplete

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/FdItem.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * FdItem
+ */
+public class FdItem {
+    private ImmutableSet<SlotReference> parentExprs;
+
+    private boolean isUnique;
+
+    private boolean isCandidate;
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/FdItem.java:33]: what's the meaning of the candidate? It seems to be not obvious

**Comment by xzj7019** [comment]
run buildall

**Comment by xzj7019** [comment]
run buildall

**Comment by xzj7019** [comment]
run buildall

**Comment by xzj7019** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupByKey.java [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+            int minParentExprCnt = -1;
+            ImmutableSet<SlotReference> minParentExprs = ImmutableSet.of();
+            // if unique fd items exists, try to find the one which has the
+            // smallest parent exprs
+            for (int i = 0; i < uniqueFdItems.size(); i++) {
+                FdItem fdItem = uniqueFdItems.get(i);
+                ImmutableSet<SlotReference> parentExprs = fdItem.getParentExprs();
+                if (minParentExprCnt == -1 || parentExprs.size() < minParentExprCnt) {
+                    boolean isContain = isExprsContainFdParent(candiExprs, fdItem);
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupByKey.java:83]: Is it redundant? If there are unique slots, we can always use one unique slot to replace all group by keys

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupByKey.java [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+            // find the root expr, add into root exprs set, indicate the index in
+            // candiExprs list
+            for (int i = 0; i < rootExprs.size(); i++) {
+                int index = findEqualExpr(candiExprs, rootExprs.get(i));
+                if (index != -1) {
+                    rootExprsSet.add(new Integer(index));
+                } else {
+                    return null;
+                }
+            }
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupByKey.java:159]: We don't need it after checking isContaining

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupByKey.java [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+                // if any unique fd item found, find the expr which matching parentExprs
+                // from candiExprs directly
+                for (int i = 0; i < minParentExprs.size(); i++) {
+                    int index = findEqualExpr(candiExprs, minParentExprs.asList().get(i));
+                    if (index != -1) {
+                        rootExprsSet.add(new Integer(index));
+                    } else {
+                        return null;
+                    }
+                }
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupByKey.java:104]: We don't need it after checking isContaining

**Review by keanji-x — APPROVED** [review]

**Pull request merged**